filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilderFactory.scala

```diff
@@ -1,27 +1,120 @@
 package cromwell.filesystems.blob
 
 import akka.actor.ActorSystem
-import com.azure.core.credential.AzureSasCredential
+import com.azure.core.management.AzureEnvironment
+import com.azure.core.management.profile.AzureProfile
+import com.azure.identity.DefaultAzureCredentialBuilder
+import com.azure.resourcemanager.AzureResourceManager
+import com.azure.storage.blob.BlobServiceClientBuilder
+import com.azure.storage.common.StorageSharedKeyCredential
+import com.azure.storage.common.sas.{AccountSasPermission, AccountSasResourceType, AccountSasService, AccountSasSignatureValues}
 import com.typesafe.config.Config
 import cromwell.core.WorkflowOptions
 import cromwell.core.path.PathBuilderFactory
-import cromwell.filesystems.blob.BlobPathBuilder
 import net.ceedubs.ficus.Ficus._
 
+import java.time.OffsetDateTime
 import scala.concurrent.ExecutionContext
 import scala.concurrent.Future
 
+import scala.jdk.CollectionConverters._
+
 final case class BlobFileSystemConfig(config: Config)
 final case class BlobPathBuilderFactory(globalConfig: Config, instanceConfig: Config, singletonConfig: BlobFileSystemConfig) extends PathBuilderFactory {
   val sasToken: String = instanceConfig.as[String]("sas-token")
   val container: String = instanceConfig.as[String]("store")
   val endpoint: String = instanceConfig.as[String]("endpoint")
   val workspaceId: String = instanceConfig.as[String]("workspace-id")
-  val workspaceManagerURL = singletonConfig.config.as[String]("workspace-manager-url")
+  val workspaceManagerURL: String = singletonConfig.config.as[String]("workspace-manager-url")
+
+  val blobTokenGenerator: BlobTokenGenerator = BlobTokenGenerator.createBlobTokenGenerator(
+    container, endpoint, Option(workspaceId), Option(workspaceManagerURL))
 
   override def withOptions(options: WorkflowOptions)(implicit as: ActorSystem, ec: ExecutionContext): Future[BlobPathBuilder] = {
     Future {
-      new BlobPathBuilder(new AzureSasCredential(sasToken), container, endpoint)
+      new BlobPathBuilder(blobTokenGenerator, container, endpoint)
+    }
+  }
+}
+
+sealed trait BlobTokenGenerator {
+  def getAccessToken: String
+}
+
+object BlobTokenGenerator {
+  def createBlobTokenGenerator(container: String, endpoint: String): BlobTokenGenerator = {
+    createBlobTokenGenerator(container, endpoint, None, None)
```

Comment:
I feel like this is partially reimplementing method overloading. You could either directly return `NativeBlobTokenGenerator(container, endpoint)` right here, or add default `None` values for the optional parameters on the other `createBlobTokenGenerator` and remove this `createBlobTokenGenerator`.

Reply:
Also, if you name the function(s) `apply`, you can call it with constructor-like syntax, e.g.
```
BlobTokenGenerator(store, endpoint)
```
instead of
```
BlobTokenGenerator.createBlobTokenGenerator(store, endpoint)
```